File: src/test/java/org/mozilla/msrp/platform/profile/FirefoxAccountServiceTest.java
package org.mozilla.msrp.platform.profile;

import org.json.JSONException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.test.context.junit4.SpringRunner;

import java.io.IOException;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.mozilla.msrp.platform.profile.FirefoxAccountService.API_AUTH_RESPONSE_KEY_ACCESS_TOKEN;

@RunWith(SpringJUnit4ClassRunner.class)
public class FirefoxAccountServiceTest {

    @Test
    public void some_meaning_less_tes() {
        final String fakeToken = "fake";
        final String fakeTokenResponse = "{" +
                "\"" + API_AUTH_RESPONSE_KEY_ACCESS_TOKEN + "\":" + "\"" + fakeToken + "\"" +
                "}";
        FirefoxAccountService service = new FirefoxAccountService();
        Retrofit retrofit = mock(Retrofit.class);
        service.setRetrofit(retrofit);
        FirefoxAccountServiceInfo info = new FirefoxAccountServiceInfo("1", "secret", fakeToken, "profile");
        service.setFirefoxAccountServiceInfo(info);
        try {

            when(retrofit.load(Mockito.any())).thenReturn(fakeTokenResponse);
            assertEquals(fakeToken, service.authorization(""));

        } catch (IOException | JSONException e) {
            e.printStackTrace();
            fail();
        }
    }
}

File: src/test/java/org/mozilla/msrp/platform/profile/FirefoxAccountServiceTest.kt
package org.mozilla.msrp.platform.profile

import okhttp3.ResponseBody
import org.junit.Test

import org.junit.Assert.*
import org.junit.Before
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.mockito.stubbing.OngoingStubbing
import org.mozilla.msrp.platform.common.mock
import org.mozilla.msrp.platform.common.onRetrofitExecute
import org.springframework.http.HttpStatus
import retrofit2.Call
import retrofit2.Response


class FirefoxAccountServiceTest {

    lateinit var firefoxAccountService: FirefoxAccountService

    @Mock
    lateinit var authClient: FirefoxAccountClient

    @Mock
    lateinit var profileClient: FirefoxAccountClient

    @Before
    fun setUP() {
        authClient = mock(FirefoxAccountClient::class.java)
        profileClient = mock(FirefoxAccountClient::class.java)
        firefoxAccountService = FirefoxAccountService(
            authClient,
            profileClient,
            FirefoxAccountServiceInfo("", "", "", "")
        )

    }

    @Test
    fun getTokenSuccess() {

        val fxaTokenRequest = firefoxAccountService.genFxaTokenRequest("")

        val accessToken = "token"

        onRetrofitExecute(authClient.token(fxaTokenRequest))
            ?.thenReturn(Response.success(FxaTokenResponse(accessToken, "")))

        val response = firefoxAccountService.token(fxaTokenRequest)

        assertTrue(response == accessToken)

    }

    @Test
    fun getTokenFail() {

        val fxaTokenRequest = firefoxAccountService.genFxaTokenRequest("")

        onRetrofitExecute(authClient.token(fxaTokenRequest))
            ?.thenReturn(Response.error(400, mock(ResponseBody::class.java)))

        val response = firefoxAccountService.token(fxaTokenRequest)

        assertTrue(response == null)

    }

    @Test
    fun getProfileSuccess() {
        val profileResponse = FxaProfileResponse()

        val bearer = "bear i-am-a-fake-bearer"

        onRetrofitExecute(profileClient.profile(bearer))
            ?.thenReturn(Response.success(profileResponse))

        val response = firefoxAccountService.profile(bearer)

        assertTrue(response == profileResponse)
    }

    @Test
    fun getProfileFail() {

        val bearer = "bear i-am-a-fake-bearer"

        onRetrofitExecute(profileClient.profile(bearer))
            ?.thenReturn(Response.error(400, mock(ResponseBody::class.java)))

        val response = firefoxAccountService.profile(bearer)

        assertTrue(response == null)
    }
}
